utils/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/source/KafkaAvroSourceFactory.scala

```diff
@@ -118,19 +114,20 @@ class KafkaAvroSourceFactory[K: ClassTag, V: ClassTag](val schemaRegistryProvide
 
   override def paramsDeterminedAfterSchema: List[Parameter] = Nil
 
-  override def implementation(params: Map[String, Any], dependencies: List[NodeDependencyValue], finalState: Option[State]): FlinkSource[ConsumerRecord[K, V]] = {
+  override def implementation(params: Map[String, Any], dependencies: List[NodeDependencyValue], finalState: Option[State]): Source[ConsumerRecord[K, V]] = {
     val preparedTopic = extractPreparedTopic(params)
     val KafkaAvroSourceFactoryState(keySchemaDataUsedInRuntime, valueSchemaUsedInRuntime, kafkaContextInitializer) = finalState.get
 
     // prepare KafkaDeserializationSchema based on given key and value schema (with schema evolution)
-    val deserializationSchema = schemaRegistryProvider.deserializationSchemaFactory.create[K, V](kafkaConfig, keySchemaDataUsedInRuntime, valueSchemaUsedInRuntime)
+    val deserializationSchema = schemaRegistryProvider
+      .deserializationSchemaFactory.create[K, V](kafkaConfig, keySchemaDataUsedInRuntime, valueSchemaUsedInRuntime)(keyClassTag, valueClassTag)
 
     // - avro payload formatter requires to format test data with writer schema, id of writer schema comes with event
     // - for json payload event does not come with writer schema id
-    val formatterSchema = schemaRegistryProvider.deserializationSchemaFactory.create[K, V](kafkaConfig, None, None)
-    val recordFormatter = schemaRegistryProvider.recordFormatterFactory.create[K, V](kafkaConfig, formatterSchema)
+    val formatterSchema = schemaRegistryProvider.deserializationSchemaFactory.create[K, V](kafkaConfig, None, None)(keyClassTag, valueClassTag)
+    val recordFormatter = schemaRegistryProvider.recordFormatterFactory.create[K, V](kafkaConfig, formatterSchema)(keyClassTag, valueClassTag)
 
-    createSource(params, dependencies, finalState, List(preparedTopic), kafkaConfig, deserializationSchema, timestampAssigner, recordFormatter, kafkaContextInitializer)
+    createSource(params, dependencies, finalState, List(preparedTopic), kafkaConfig, deserializationSchema, recordFormatter, kafkaContextInitializer)
```

Comment:
I wonder, do we need inheritance here? It seems that everything is passed in arguments?

Reply:
You mean by replacing it with composition? Yes, I am a great fun of this. But have no idea how to make it not hurting too much during migration. New parameter like `KafkaAvroSourceImplementationFactory`? It looks weird because this new extracted class is a real factory and all the code in `KafkaAvroSourceFactory` is kind of `KafkaAvroSourceFactoryDefinitionAndImplementationFacade` :/. Any idea?

Reply:
Yes, sth like ImplementationFactory, the same with Sink (see comment in KafkaAvroSinkFactory). I think we made so many changes that one more won't hurt that much. What's more, I think that Avro components are less frequently customised/changed than 'generic' (where e.g. deployments use their own model)

Reply:
The true benefit was only when I've change it in both KafkaAvroSourceFactory and KafkaSourceFactory - changed. A lot of duplicated code disappear - take a look. Tomorrow I will do the same with Sinks